velox/exec/TaskDriverOperatorLifecycle.md

```diff
@@ -0,0 +1,96 @@
+# Task, Driver, Operator lifecycle
+
+__TL;DR__ Task outlives Drivers and Operators. There are circular references between
+Tasks and Drivers which are cleared when Task drops references to Drivers.
+Task::terminate contains catch-all logic to clear things up in case of abnormal
+or early termination.
+
+## PartitionedOutputBufferManager
+
+PartitionedOutputBuffer holds a reference to the task. This reference is used to
+notify the task that all output has been consumed by calling
+task->setAllOutputConsumed() from PartitionedOutputBuffer::deleteResults.
+
+This reference is dropped when PartitionedOutputBuffer is destroyed which
+happens in (1) PartitionedOutputBufferManager::deleteResults (happy path) and
+(2) PartitionedOutputBufferManager::removeTask (error path).
+
+In Prestissimo application, PartitionedOutputBufferManager::deleteResults is called by
+TaskManager::abortResults which is called by TaskResource::abortResults
+attached to HTTP DELETE /v1/task/(.+)/results/(.+)
+
+PartitionedOutputBufferManager::removeTask is called by Task::terminate.
+
+## Operator
+
+Operator doesn’t hold a reference to Task directly. It has access to it via
+DriverCtx.
+
+## Driver and Task
+
+DriverCtx stores a reference to a Task. Driver has an exclusive ownership of
+DriverCtx and has a reference to a task via DriverCtx. Operators reference
+DriverCtx via a raw pointer.
+
+Driver:
+
+    std::unique_ptr<DriverCtx> ctx_;
+
+DriverCtx:
+
+    std::shared_ptr<Task> task;
+
+Driver’s reference to a Task is dropped only on Driver’s destruction.
+
+Task has a shared ownership of Drivers.
+
+    std::vector<std::shared_ptr<Driver>> drivers_;
+
+Task stores references to Drivers in two places: (1) drivers_ member variable
+that stores all drivers; (2) SplitGroupState.barriers stores references to
+Drivers containing join builds (these are used in Task::allPeersFinished).
+
+References to drivers_ are cleared in (1) Task::removeDriver called from
+Driver::close; (2) Task::terminate.
+
+References to Drivers stored in barriers are cleared in (1) Task::removeDriver
+via SplitGroupState::clear(); (2) Task::allPeersFinished when all join build
+pipelines finish; (3) Task::terminate via SplitGroupState::clear().
+
+Task::removeDriver and Task::allPeersFinished are called in happy paths
+(including a path where Driver itself encounters an error and closes itself).
+Task::terminate is called in the error path.
+
+There are circular references between Tasks and Drivers where Task references
+Drivers and Drivers reference Tasks. These circular references are cleared when
+Task drops references to Drivers. Driver holds at least one reference
```

Comment:
These sentences are distracting especially if all references between tasks and drivers have been covered between lines 30-62. Just giving the text from line 66... "Since Driver holds at least one reference ... until destruction..."

Reply:
I'm not sure I understand. I wanted to clarify that there are indeed circular references. A careful reader would notice it already, but they may wonder whether this is indeed the case.